llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -7865,6 +7867,65 @@ LegalizerHelper::lowerVectorReduction(MachineInstr &MI) {
   return UnableToLegalize;
 }
 
+static Type *getTypeForLLT(LLT Ty, LLVMContext &C);
+
+LegalizerHelper::LegalizeResult LegalizerHelper::lowerVAArg(MachineInstr &MI) {
+  Observer.changingInstr(MI);
+  MachineFunction &MF = *MI.getMF();
+  const DataLayout &DL = MIRBuilder.getDataLayout();
+  LLVMContext &Ctx = MF.getFunction().getContext();
+  Register ListPtr = MI.getOperand(1).getReg();
+  LLT PtrTy = MRI.getType(ListPtr);
+
+  // LstPtr is a pointer to the head of the list. Get the address
+  // of the head of the list.
+  Align PtrAlignment = Align(DL.getABITypeAlign(getTypeForLLT(PtrTy, Ctx)));
+  MachineMemOperand *PtrLoadMMO =
+      MF.getMachineMemOperand(MachinePointerInfo::getUnknownStack(MF),
+                              MachineMemOperand::MOLoad, PtrTy, PtrAlignment);
+  auto HeadOfList = MIRBuilder.buildLoad(PtrTy, ListPtr, *PtrLoadMMO).getReg(0);
+  Register VAList = HeadOfList;
+
+  const MaybeAlign MA(MI.getOperand(2).getImm());
```

Comment:
I think you can use Align instead of MaybeAlign based on how `IRTranslator::translateVAArg` works. It calls `getABITypeAlign` which returns an `Align` and then converts that to an integer.

Reply:
I might have been wrong about that. It looks like the type legalizer in SelectionDAG sets the alignment to 0 to indicate when a ISD::VAARG has been split. We don't have narrowScalar support for G_VAARG yet so we haven't encountered the need for that yet.

Given the way it's used, an alignment of 1 would probably have worked equally well, but maybe 0 made it more obvious to anyone debugging?

Reply:
> IRTranslator::translateVAArg uses getABITypeAlign which returns an Align and then converts that to an integer.

We're always going to get an `Align`, not a MaybeAlign. I think we can use Align. SDAG does use 0 to indicate the Hi part of the split, but we are not doing that today in GISel, so I don't think it matters?

Reply:
I've added `clampScalar` to this patch which will return UnableToLegalize for now. We will need to implement splitting in a future patch. I think we can use an Align for now and defer the decision on whether it needs to be an Align or Maybe align when there is clampScalar support which is playing with the Align values.